### PR TITLE
Remove more IsDisabled checks

### PR DESCRIPTION
**Pull request by reaperrr**
Depends on #13998.

This removes most of the remaining `Actor.IsDisabled` look-ups.
After this, only `Gate`,  `ProductionQueue` and AI `AirStates` remain (might be worth investigating whether those are easy enough to migrate to add them to this PR, too).

### CODE REVIEW

**Comment by GraionDilach** [comment]
Wouldn't https://github.com/OpenRA/OpenRA/pull/14202/commits/c7b7d8f874e3eb2c6cf69ae051fb3408296a53cf conflict with the ammo refactors?

**Comment by pchote** [comment]
> OpenRA.Mods.Common/Traits/AffectsShroud.cs(70,19): error CS0103: The name `IsDisabled' does not exist in the current context
> 
> OpenRA.Mods.Common/Traits/AffectsShroud.cs(73,87): error CS0103: The name `cachedDisabled' does not exist in the current context
> 
> OpenRA.Mods.Common/Traits/AffectsShroud.cs(105,38): error CS0103: The name `cachedDisabled' does not exist in the current context
> 
> OpenRA.Mods.Common/Traits/Turreted.cs(99,13): error CS0103: The name `self' does not exist in the current context
> 
> Compilation failed: 4 error(s), 0 warnings

File: OpenRA.Mods.Common/Traits/Turreted.cs [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
 namespace OpenRA.Mods.Common.Traits
 {
-	public class TurretedInfo : ITraitInfo, UsesInit<TurretFacingInit>, Requires<BodyOrientationInfo>, IActorPreviewInitInfo
+	public class TurretedInfo : ITraitInfo, UsesInit<TurretFacingInit>, Requires<BodyOrientationInfo>, Requires<AttackTurretedInfo>, IActorPreviewInitInfo
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Turreted.cs]: I really don't like the idea of making this depend on `AttackTurreted`.  Can we do this a different way?

File: OpenRA.Mods.Common/Traits/Attack/AttackBase.cs [hunk 2]
```diff
@@ -81,7 +81,7 @@ protected virtual Func<IEnumerable<Armament>> InitializeGetArmaments(Actor self)
 
 		protected virtual bool CanAttack(Actor self, Target target)
 		{
-			if (!self.IsInWorld || IsTraitDisabled || self.IsDisabled())
+			if (!self.IsInWorld || IsTraitDisabled)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]: Shouldn't this return false on `IsTraitPaused` too?

**Comment by reaperrr** [comment]
Updated and rebased.

File: OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs [hunk 3]
```diff
@@ -84,7 +84,7 @@ public override Activity Tick(Actor self)
 				if (IsCanceled || !target.IsValidFor(self))
 					return NextActivity;
 
-				if (self.IsDisabled())
+				if (attack.IsTraitDisabled)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs]: AttackBase maps `IsDisabled()` &rarr; `IsTraitPaused`, so shouldn't that be the same here?

File: OpenRA.Mods.Common/Traits/Attack/AttackTurreted.cs [hunk 4]
```diff
@@ -17,17 +17,22 @@ namespace OpenRA.Mods.Common.Traits
...
 	{
+		[Desc("Turret names")]
+		public readonly string[] Turrets = { "primary" };
+
 		public override object Create(ActorInitializer init) { return new AttackTurreted(init.Self, this); }
 	}
 
 	public class AttackTurreted : AttackFollow
 	{
+		public readonly AttackTurretedInfo AttackTurretedInfo;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Attack/AttackTurreted.cs]: This doesn't look very nice.  Just cast it when you need it instead?

File: OpenRA.Mods.Common/Traits/Turreted.cs [hunk 5]
```diff
@@ -113,38 +114,30 @@ void ITick.Tick(Actor self)
...
-						realignTick++;
-					else if (info.RealignDelay > -1)
-						DesiredFacing = null;
-
-					MoveTurret();
-				}
-			}
-			else
-			{
-				realignTick = 0;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Turreted.cs]: What happened to this line?

File: OpenRA.Mods.Common/Traits/Attack/AttackTurreted.cs (hunk 4 above)
Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/Attack/AttackTurreted.cs]: Not sure what you mean with casting?
I couldn't come up with anything else that compiled and allowed looking up this AttackTurreted-specific Info in Turreted.

**Comment by reaperrr** [comment]
Updated.

**Comment by pchote** [comment]
Regression: The RA gap generator does not disable on low power.

**Comment by pchote** [comment]
Regression: RA tesla coils (and other powered defenses) lose the ability to target and their range circles when powered down.  Something seems to be treating them as disabled rather than paused.

**Comment by reaperrr** [comment]
> Regression: The RA gap generator does not disable on low power.
> Regression: RA tesla coils (and other powered defenses) lose the ability to target and their range circles when powered down. Something still seems to be treating them as disabled rather than paused.

Fixed. These were introduced by #13998, but at least some of them needed Attack* traits to be pausable to fix them, so this is the right place to fix them anyway.

**Comment by pchote** [comment]
I'm not sure if this is a new regression or and old bug, but if you build an AA gun (or another turreted defense), order it to force attack a Mig (or another unit) and then immediately disable it then the turret will continue to track while disabled, without firing.

**Comment by pchote** [comment]
Something else that i'm not sure about: in some cases disabling an Attack* seems to cancel the target, and in others not.

**Comment by GraionDilach** [comment]
The turrets-stuck-in-attack mode bug (which the last two instances have showcased) have been already filed as #12856.

**Comment by reaperrr** [comment]
> Something else that i'm not sure about: in some cases disabling an Attack* seems to cancel the target, and in others not.

I assume *disabling* should always cancel it?

**Comment by pchote** [comment]
Sorry, I meant disabling in the sense of EMP or power-down ingame. i.e. Pausing.

**Comment by reaperrr** [comment]
> Edit: This is indeed a regression vs the last release.
Edit: this is also a regression vs the last release.

This doesn't make any sense to me. Going by code changes, I see no way how `PauseOnCondition` should behave any different to Actor.IsDisabled on last release.

Both 'regressions' should already be present on last release, going by the code changes in this PR alone.

**Comment by reaperrr** [comment]
Ok, managed to find the cause of the first one. Still investigating the 2nd one.

**Comment by reaperrr** [comment]
Updated.

File: mods/ts/rules/gdi-vehicles.yaml [hunk 6]
```diff
@@ -364,7 +369,7 @@ JUGG:
 	AttackTurreted@deployed:
 		Voice: Attack
 		Armaments: deployed
-		RequiresCondition: deployed
+		PauseOnCondition: empdisable || !deployed
```
Comment (pchote) [inline mods/ts/rules/gdi-vehicles.yaml]: Shouldn't this be
```
RequiresCondition: deployed
PauseOnCondition: empdisable
```
?

File: mods/ts/rules/nod-vehicles.yaml [hunk 7]
```diff
@@ -138,7 +140,7 @@ TTNK:
 		RequiresCondition: !undeployed && real-actor
 	AttackFrontal:
 		Voice: Attack
-		RequiresCondition: undeployed
+		RequiresCondition: !empdisable && undeployed
```
Comment (pchote) [inline mods/ts/rules/nod-vehicles.yaml]: Shouldn't these be 
```
RequiresCondition: undeployed
PauseOnCondition: empdisable
```
and
```
RequiresCondition: deployed
PauseOnCondition: empdisable
```
?  We do want to switch these off completely when in the opposite deploy state.

File: mods/ts/rules/nod-vehicles.yaml [hunk 8]
```diff
@@ -251,7 +253,7 @@ ART2:
 	AttackTurreted@deployed:
 		Voice: Attack
 		Armaments: deployed
-		RequiresCondition: deployed
+		PauseOnCondition: empdisable || undeployed
```
Comment (pchote) [inline mods/ts/rules/nod-vehicles.yaml]: Likewise here.

**Review by pchote — COMMENTED** [review]
The latest version seems to have fixed all the regressions in the classic mods, great job!

I found a couple of possible issues in TS:
* Deployed Tick Tanks and Juggernauts can't attack.
* Inline comments below

**Comment by reaperrr** [comment]
Updated.

> Deployed Tick Tanks and Juggernauts can't attack.

Found and fixed the problem. `AttackTurreted`'s new `Turrets` property defaults to "primary", while the turrets of those 2 + arty had their name changed to `deployed`.

**Review by pchote — APPROVED** [review]

**Comment by reaperrr** [comment]
@abcdefg30 @penev92 There's one last follow-up waiting that will remove the remaining 3 uses of `Actor.IsDisabled()`, so this is also one of the PRs that should get merged soon so we can finally wrap this up.

**Review by penev92 — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/babcab0fd80d2d73a4f3565ed0caad6fe7cd55ef...3441324c2160290ca56f10257a8b3f1827e80012

**Pull request merged**